File: Yr2022/day_twentythree.py
from time import perf_counter_ns
from tqdm import tqdm

from itertools import groupby
from operator import countOf
from typing import Tuple, Set, List, Dict

from dataclasses import dataclass

UPBITS = 0b11100000
DOWNBITS = 0b00000111
LEFTBITS = 0b10010100
RIGHTBITS = 0b00101001

@dataclass
class Arena:
    occupied: Set[Tuple[int,int]]
    adjs = [[UPBITS,(0,-1)], [DOWNBITS,(0,1)], [LEFTBITS,(-1,0)], [RIGHTBITS,(1,0)]]
    def get_adjacents(self, pos):
        delta_x, delta_y = [1, 0, -1], [1, 0, -1]
        idx, bits = 0, 0
        for dy in delta_y:
            for dx in delta_x:
                if dx ==  0 and dy == 0:
                    continue
                if (pos[0] + dx, pos[1] + dy) in self.occupied:
                    bits = bits | (1 << idx)
                idx += 1

        return bits
    
    def proposals(self, pos, turn_idx):
        adj = self.get_adjacents(pos)
        if adj == 0:            
            return None
        for _i in range(0,4):
            side_bits = self.adjs[ (turn_idx+_i) % len(self.adjs)][0]
            side_dir = self.adjs[ (turn_idx+_i) % len(self.adjs)][1]
            side_clear = adj & side_bits == 0
            if side_clear:
                return (pos[0]+side_dir[0], pos[1]+side_dir[1])
        return None

    def calc_move(self, turn_idx):
        move_dict = {}
        next_occupied = set()
        for current in self.occupied:
            move_dict[current] = self.proposals(current, turn_idx)
      
        for current, next in move_dict.items():
            
            if next is None or countOf(move_dict.values(), next) > 1:
                next_occupied.add(current)
            else:
                next_occupied.add(next)
        has_updated = self.occupied != next_occupied
        self.occupied = next_occupied
        return has_updated
    
    def get_empty(self):
        xs, ys = [pos[0] for pos in self.occupied], [pos[1] for pos in self.occupied]
        min_x, max_x = min(xs), max(xs)+1
        min_y, max_y = min(ys), max(ys)+1
        total_area = (max_y-min_y)* (max_x - min_x)    
        
        return total_area - len(self.occupied)
        

    def print_arena(self):
        xs, ys = [pos[0] for pos in self.occupied], [pos[1] for pos in self.occupied]
        min_x, max_x = min(xs + [0]), max(xs + [4])
        min_y, max_y = min(ys + [0]), max(ys + [5])
        out_str = ""
        for y in range(min_y, max_y+1):
            for x in range(min_x, max_x+1):
                out_str += "#" if (x,y) in self.occupied else "."
            out_str += "\n"
        print(out_str)


        
def pathfind(filename: str):    
    with open(filename, 'r') as file:
        cmd_lines = file.readlines()
    
    occupied = set()
    for line_idx, line in enumerate(cmd_lines):
        for char_idx, char in enumerate(line):
            if char == '#':
                occupied.add( (char_idx, line_idx ))

    arena = Arena(occupied=occupied)
    #arena.print_arena()
    for i in range(0, 10):        
        arena.calc_move(i)        
       
    
    part_one = arena.get_empty()
    part_two, has_updated = 0, True
    arena_two = Arena(occupied=occupied)
    while has_updated:
        
        has_updated = arena_two.calc_move(part_two)
        part_two += 1     

    return part_one, part_two
  

if __name__ == "__main__":
    time_start = perf_counter_ns()
    file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\puzzle_day_twentythree.txt"
    #file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\example.txt"
    #file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\other_example.txt"
    
    part_one, part_two = pathfind(file_loc)
    
    print(f"(Part One): {part_one}     (Part Two): {part_two}")   
    print(f"Elapsed time : {(perf_counter_ns() - time_start)/1000:.0f} microseconds")  

File: Yr2022/src/day_twentythree.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;
use std::collections::{HashSet, HashMap};

type Pos = (i32,i32);



fn main() {   
    let start = Instant::now();
    let file_name= "C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\puzzle_day_twentythree.txt";
    //let file_name= "C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\example.txt";
    let (part_one, part_two) = path_find(file_name);
    println!("Part One: {}  Part Two:  {}", part_one, part_two);
    let duration = start.elapsed();    
    println!("Time taken to execute: {} microseconds", duration.as_micros());    
}

struct AdjDir {
    bits: u8,
    dx: i32,
    dy: i32,
}

struct Arena {
    occupied: HashSet<Pos>,
    adjs: Vec<AdjDir>,
}

impl Arena {
    fn get_adjacents(&self, pos: Pos) -> u8 {
        let mut bits: u8 = 0;
        for (idx, (dx,dy)) in vec![(1,1), (0,1), (-1,1), (1,0), (-1,0), (1,-1), (0,-1), (-1,-1)].iter().enumerate() {
            if self.occupied.contains(&(pos.0 + dx, pos.1 + dy)) {
                bits = bits | (1<<idx);
            }
        }
        return bits;
    }

    fn proposals(&self, pos: Pos, turn_idx: usize) -> Option<Pos> {
        let adj = self.get_adjacents(pos);
        if adj == 0 { return None; }
        for i in 0..4 {
            let adj_dir = &self.adjs[ (turn_idx+i)% self.adjs.len() ];
            if adj_dir.bits & adj == 0 { return Some((pos.0+adj_dir.dx, pos.1+adj_dir.dy));}
        }
        return None;
    }

    fn calc_move(&mut self, turn_idx: usize) -> bool {
        let mut move_dict : HashMap<Pos, Option<Pos>> = HashMap::new();
        let mut next_occupied : HashSet<Pos> = HashSet::new();
        let mut has_updated: bool = false;
        for current in &self.occupied {
            move_dict.insert(*current, self.proposals(*current, turn_idx));
        }

        for (current, next) in move_dict.iter() {
            if next.is_none() {next_occupied.insert(*current);}
            else {
                let next_count = move_dict.values().filter(|v| 
                    match v {
                        Some(entry) => *entry == next.unwrap(),
                        _ => false,
                    }
                ).count();           
                
                if next_count > 1 {
                    next_occupied.insert(*current);
                }
                else {
                    next_occupied.insert(next.unwrap());
                    has_updated = true;
                }
            }
        }
        
        self.occupied = next_occupied;
        return has_updated;
    }

    fn get_empty(&self) -> i32 {
        let min_x = self.occupied.iter().map(|p| p.0).min().unwrap();
        let max_x = self.occupied.iter().map(|p| p.0).max().unwrap()+1;
        let min_y = self.occupied.iter().map(|p| p.1).min().unwrap();
        let max_y = self.occupied.iter().map(|p| p.1).max().unwrap()+1;
        return (max_y-min_y)*(max_x-min_x) - self.occupied.len() as i32;
    }

    fn print_arena(&self) {
        let min_x = self.occupied.iter().map(|p| p.0).min().unwrap();
        let max_x = self.occupied.iter().map(|p| p.0).max().unwrap();
        let min_y = self.occupied.iter().map(|p| p.1).min().unwrap();
        let max_y = self.occupied.iter().map(|p| p.1).max().unwrap();
        let mut out_str = String::new();
        for y in min_y..=max_y{
            for x in min_x..=max_x {
                if self.occupied.contains(&(x,y)) {
                    out_str.push('#');
                }
                else {
                    out_str.push('.');
                }
            }
            out_str.push('\n');
        }
        println!("{}", out_str);
    }
}



fn path_find(my_file_name: &str) -> (i32, usize){        
    let file = File::open(my_file_name).expect("File does not exist");
    let buf_lines = BufReader::new(file).lines();
    let line_vec: Vec<String> = buf_lines.map(|l| l.expect("Fail to parse")).collect();

    let mut occupied : HashSet<Pos> = HashSet::new();
    for (line_idx, line) in line_vec.iter().enumerate() {
        for char_idx in 0..line.len() {
            if line.chars().nth(char_idx) == Some('#') {
                occupied.insert( (char_idx as i32, line_idx as i32) );
            }
        }
    }
    let up_bits = 0b11100000;
    let down_bits = 0b00000111;
    let left_bits = 0b10010100;
    let right_bits = 0b00101001;
    let adjs = vec![AdjDir{bits:up_bits, dx:0, dy:-1},
                AdjDir{bits:down_bits, dx:0, dy:1},
                AdjDir{bits:left_bits, dx:-1, dy:0},
                AdjDir{bits:right_bits, dx:1, dy:0}];
    let mut arena = Arena{occupied:occupied, adjs:adjs };

    let mut has_updated: bool = true;
    let mut turn_idx: usize = 0;
    let mut part_one = 0;

    if false {
        arena.print_arena();
    }

    while has_updated {
        if turn_idx == 10 {part_one = arena.get_empty();}       
        has_updated = arena.calc_move(turn_idx);
        turn_idx += 1;
    }
    let part_two = turn_idx;  
    
    
    return (part_one, part_two);
}


